### ISSUE TITLE
docs: document `table_by_pk` query

### ISSUE THREAD

**Issue by corysimmons** [hasura/graphql-engine#2047]
Instead of `query { person(where: ... ) { id } }` returning an array of `person`, it would be nice if it returned a single object.

This pattern of duplicating interactions for both plural (arrays) and singular (objects) results can be applied to mutations as well as queries.

Example:

```js
// before
{
  users(where: {
    id: {
      _eq: 123
    }
  }) {
    id
  }
} // returns an array of users even if I just want one... this makes destructuring and working with these values messy

// after
{
  user (where: {
    id: {
      _eq: 123
    }
  }) {
    id
  }
} // returns a single user object

// OR

{
  users (where: {
    id: {
      _eq: 123
    }
  }) {
    id
  }
} // returns an array of users that might match that query

// before
mutation AddUser ($objects: [users_insert_input!]!) { // I only want to insert a single user but I have to submit the mutation as an array.
  insert_users(objects: $objects) {
    returning {
      id
    }
  }
}

// after
mutation AddUser ($object: user_insert_input!) { // only inserting a single user; returning a single object
  insert_user(object: $object) {
    returning {
      id
    }
  }
}

// OR

mutation AddUsers ($objects: [users_insert_input!]!) { // Sometimes I will want to add multiple users, so keep the existing queries/mutations available.
  insert_users(objects: $objects) {
    returning {
      id
    }
  }
}
```

GraphCMS does this and it's a really nice user experience. In contrast, this is really one of the few things I find actually hinders my productivity and ability to read the code with Hasura.

**Comment by revskill10** [comment]
@corysimmons I think, adding an argument like `{singular: true}` (default is false) is enough ?

**Comment by corysimmons** [comment]
I don't think it requires clientside anything and wouldn't be a breaking change?

`users_insert_input` gets turned into

- `users_insert_input`
- and `user_insert_input`

Adding `{ singular: true }` to everything would be better than nothing, but isn't quite as elegant as just using singular named queries/mutations.

**Comment by corepay** [comment]
@corysimmons running into similar issue now and yes returning an object on a query would be nice so I can reuse components without restructuring.    Question though (i am still new at all this..) is my table names are singular now - - - so my single object query (for example `for` returning an array (causing me grief) are singular...  

```
{
query fetchForm($id: String!) {
  form(where: {id: {_eq: $id}}) {
    id
   ...
  }
}
```

Are you suggesting that I should be able to use `forms(where:...` and Hasura will recognize this is an array request but will remap the current singular table query `form(where:...` as an object response?

**Comment by corysimmons** [comment]
Yeah, I suppose Hasura would have to have to enforce your model being singular in the db, and then Hasura would just intelligently provide queryable/mutatable plurals and singulars. 

fwiw, currently I'm going against my gut and naming db models plural just so the querying/mutating doesn't feel so off...

**Comment by OLDIN** [comment]
When will support for a single query be implemented?

**Comment by coco98** [comment]
@OLDIN There's already a `table_by_pk` style query root that you should see. We're going to make it easier to rename that to a better form.

**Comment by revskill10** [comment]
@coco98 I see Hasura understands unique key constraints, so is it possible to autogenerate `table_by_unique_key_constraint` which return singular response also ?

**Comment by corepay** [comment]
afaik - response is an array even for single object fetch. I  just extract the index[0] and assign to an object when received, problem solved.  example.  var item = this.items[0]

**Comment by corysimmons** [comment]
@corepay Yes. I know how to do that. It's annoying and unintuitive vs stuff like Prisma/GraphCMS where there is a standard that you name a model singular, then Prisma generates both singular and plural resolvers so you can query for either:

```graphql
{
  user(id: "abc") {
    first_name
  }
}
// const { first_name } = user

{
  users {
    first_name
  }
}
// users.map(u => u.first_name)
```

It's a QoL thing.

Hasura is more powerful/flexible/better than Prisma. This is the one thing I've noticed from extensively working with both that Prisma does much better than Hasura.

**Comment by OLDIN** [comment]
@coco98 I know about searching by primary key, but what about searching through other columns for a single query?

**Comment by bkstorm** [comment]
> @coco98 I know about searching by primary key, but what about searching through other columns for a single query?

yeah, for example, email is unique in user table, so I want to search by email and return only single row.

**Comment by sarendsen** [comment]
would love to see this feature

**Comment by sarendsen** [comment]
Also, seems like a duplicate to https://github.com/hasura/graphql-engine/issues/1910

**Comment by marionschleifer** [comment]
Single item selects are already available for querying. I will update the docs on this soon 🙂

**Comment by OLDIN** [comment]
@marionschleifer This is fantastic but how long would that take?

**Comment by corysimmons** [comment]
@marionschleifer Awesome! Renaming from `user_by_pk` to `user` in https://github.com/hasura/graphql-engine/pull/2509 will be amazing as well!

Thanks so much for the Hasura team's work on this!

**Comment by marionschleifer** [comment]
@OLDIN we will release this in the next few weeks 🙂

**Comment by revskill10** [comment]
Please keep the option to be optional, as current projects are still using the `user_by_pk` . My suggestion is keeping the current as default, and allow user to provide custom name if they want.

**Comment by marionschleifer** [comment]
@revskill10 it will be optional 🙂

**Comment by dukedave** [comment]
I just want to call out that this will be very useful for a specific use case of views, which helps with auditing changes to data:

1. Have a `widget_revision` table, which is insert only, and has columns for `widget_id` and `created_at` (along with other widget data).
1. When a widget is updated, the client inserts a new row in to `widget_revision`, and in Hasura we set its `created_at` to the current time. 
1. We have a `widget` view which has the most recent `widget_revision` for each `widget_id`.

It's a great pattern, but we struggle because we'd like to have `widget_by_pk`, but no such query exists. Presumably because we can't make `widget_id` on `widget` a PK, because [Postgres doesn't allow PKs on view](https://stackoverflow.com/questions/11667508/how-to-add-primary-key-to-view).

I'm **hoping** this issue would allow us to add `widget_by_pk`?

**Comment by haaah** [comment]
adding support of returning a single object will be useful in "view" scenario.

**Comment by tirumaraiselvan** [comment]
Dup of https://github.com/hasura/graphql-engine/issues/519 ?

**Comment by SidhNor** [comment]
@marionschleifer I believe the feature request wasn't just about querying by the primary key, but rather by unique columns (including primary key)
Documenting `table_by_pk` will not actually solve the initial request.

Thank you

**Comment by marionschleifer** [comment]
@SidhNor we tracked this in the following issue: https://github.com/hasura/graphql-engine/issues/1910 and implemented it with this PR: https://github.com/hasura/graphql-engine/pull/2509. We left this issue open  to document `table_by_pk`.

**Comment by marionschleifer** [comment]
@SidhNor we are tracking this feature here: https://github.com/hasura/graphql-engine/issues/519. But we left this issue open to document `table_by_pk`.

**Comment by tafelito** [comment]
@marionschleifer it seems to me that the original request was not only about changing the default name for the query_by_pk queries, but also be able to return a single object as the result of a mutation. Is that something that you are planning on add later on? 

I'm referring to this

```gql
// before
mutation AddUser ($objects: [users_insert_input!]!) { // I only want to insert a single user but I have to submit the mutation as an array.
  insert_users(objects: $objects) {
    returning {
      id
    }
  }
}

// after
mutation AddUser ($object: user_insert_input!) { // only inserting a single user; returning a single object
  insert_user(object: $object) {
    returning {
      id
    }
  }
}

// OR

mutation AddUsers ($objects: [users_insert_input!]!) { // Sometimes I will want to add multiple users, so keep the existing queries/mutations available.
  insert_users(objects: $objects) {
    returning {
      id
    }
  }
}
```

**Comment by coco98** [comment]
@tafelito Yep, single element mutations (inserts/updates/deletes) are already available in 1.2-beta* and will be available in 1.2! They implement that RFC :)

**Comment by tafelito** [comment]
Awesome thanks @coco98!

what about returning a single row for custom functions, instead of an array? I tried to find a related issue but I couldn't find it
